### PR TITLE
Ensure correct usage of language-specific object files

### PR DESCRIPTION
**Pull request by t-8ch**
This ensures the necessary rebuilds are done when building with a
different lang variable.

Previously changing the lang variable would only trigger a relink of the
final artifact, not a recompilation of all files actually changing their
behavior based on the lang parameter. Therefore the wrong language would
be used inside this firmware.

More files are recompiled than strictly necessary, but this is way is
easy and correct.

Example:

Without this PR:
```sh
$ make lang=EN
... lots of compilation
Linking Pinecil_EN.elf
$ make lang=DE  # no rebuild being done
Building for Pine64 
 
Linking Pinecil_DE.elf
riscv64-elf-objcopy Hexfile/Pinecil_DE.elf -O ihex Hexfile/Pinecil_DE.hex
riscv64-elf-size  --format=berkeley  Hexfile/Pinecil_DE.elf
   text	   data	    bss	    dec	    hex	filename
  48426	    608	  15380	  64414	   fb9e	Hexfile/Pinecil_DE.elf
riscv64-elf-objcopy Hexfile/Pinecil_DE.elf -O binary Hexfile/Pinecil_DE.bin
```

### CODE REVIEW

**Comment by Ralim** [comment]
Ah,

Actually I was thinking about this last night; 
In the build script we just jump objects between runs. But I was thinking it may be simpler to just generate a translation CPP file per language (and therefore most wouldn't contribute due to ifdefs).

Would this cause the correct behaviour instead of doing this?

**Comment by t-8ch** [comment]
This would work for `Translation.cpp` but not for users of `unit.h`.
Instead of generating a language-specific object file per user of `unit.h` I opted to just generate all object files in a language-specific way.
The per-language object directory was a very elegant solution for this.

**Comment by Ralim** [comment]
I would prefer to not have to rebuild every single object for each language if we can however to keep build times down.

**Comment by t-8ch** [comment]
Everything doing having any preprocessor macros depending on the `LANG_` codes (especially users of `unit.h`) will have to be rebuilt definitively.
I don't see a way of doing this properly.

We could have some sort of text file that contains the last language code and triggers rebuilds of `Translation.cpp` and `unit.h` if the language code is changed.
Disadvantages:
* This would also trigger a run of `make_translations.py` which is very expensive.
* This would not preserve the old object files from the previous language.

We could also manually classify all files into ones depending on translation defines or not. But that seems horrible to maintain and may lead to subtle inconsistencies.

Having per-language subdirectories would still cache all object files for a normal development-cycle rebuild, so would be as fast as previously.
Users using `build.sh` always did a full rebuild anyways.

**Comment by alvinhochun** [comment]
`unit.h` only defines the Fahrenheit support toggle. Why is this necessary instead of just enabling Fahrenheit support for all? It's a run-time option anyway.

**Comment by Ralim** [comment]
>Everything doing having any preprocessor macros depending on the LANG_ codes (especially users of unit.h) will have to be rebuilt definitively.
I don't see a way of doing this properly.

If I understand `make` correctly (which i probably don't). I think it uses file modified dates to figure out the required recompile stack. If we just touched `unit.h` and the language file before every build; shouldn't it handle the dependencies for us?


>We could also manually classify all files into ones depending on translation defines or not. But that seems horrible to maintain and may lead to subtle inconsistencies.

Yep absolutely 100% do not want that path.

>Having per-language subdirectories would still cache all object files for a normal development-cycle rebuild, so would be as fast as previously.

Agreed.

Main thing I care about build times for is the CI steps, for a single user its not really a big issue.

Would love if we could gain some speed improvement, but if thats not going to be the case; happy to go with this solution.

**Comment by t-8ch** [comment]
@alvinhochun That would be possible. My goal was to change as little as possible.

> If I understand `make` correctly (which i probably don't). I think it uses file modified dates to figure out the required recompile stack. If we just touched `unit.h` and the language file before every build; shouldn't it handle the dependencies for us?

You are right. But currently all files depend on `unit.h`, because otherwise we would have to manually figure out again where to add the dependencies.

> Main thing I care about build times for is the CI steps, for a single user its not really a big issue.

As @alvinhochun mentioned we can get rid of `unit.h` completely and export a new symbol (`have_fahrenheit` or something) from `Translation.h`/`Translation.cpp`.
Then we can compile all object files once and then compile `Translation.$(lang).o` for each language.
This way all compilation steps except `Translation.o` and the final linking can be shared between all languages.

**Comment by t-8ch** [comment]
@Ralim @alvinhochun 
I pushed a new commit that reuses all object files between different language builds (except the one from `Translation.cpp`).
For details see the commit message.

If the drawbacks are acceptable I would like to extend this even more by generating one Translation.cpp file per language directly by `make_translations.py`.
This will
* speed up compilation times.
* simplify the Makefile.
* make the generated source files easier to read.
* make it possible to build all firmwares (for a given model) fully parallelized.

File: source/Core/Inc/Settings.h [hunk 1]
```diff
@@ -9,7 +9,6 @@
 
 #ifndef SETTINGS_H_
 #define SETTINGS_H_
-#include "unit.h"
 #include <stdint.h>
 #define SETTINGSVERSION (0x24)
```
Comment (Ralim) [inline source/Core/Inc/Settings.h:13]: Can you bump this as you have changed the struct structure please ❤️

**Review by Ralim — CHANGES_REQUESTED** [review]

**Comment by Ralim** [comment]
Looks good, I'm a fan.
One trivial change so I don't forget and i'm happy to merge this in :)

**Comment by t-8ch** [comment]
@Ralim  I implemented the promised changes.
This is now capable of building all firmwares for a given model in parallel while sharing all intermediates that can be shared.
For each language a dedicated Translation.cpp file is generated.
Usage:
```bash
$ make  # Builds the default english firmware
$ make default  # same as above
$ make firmware-DE  # builds the german firmwae
$ make all  # builds all languages at once
```
Any subset of languages can be built at once by specifying all `firmware-$LANG` targets.
`build.sh` could be replaced by `make clean all` for each model.
I recommend not to execute `make` with a plain `-j` argument as it may slow down the build and instead specify an explicit value.

This now also obsoletes #807 .

File: source/Makefile [hunk 2]
```diff
@@ -126,14 +123,11 @@ OUTPUT_DIR=Objects/$(model)
 OPTIM=-Os -flto -finline-small-functions -findirect-inlining -fdiagnostics-color -ffunction-sections -fdata-sections
 
 # global defines ---------------------------------------------------------------
-GLOBAL_DEFINES += $(DEV_GLOBAL_DEFS) -D USE_RTOS_SYSTICK -D LANG_$(lang) -D LANG -D MODEL_$(model) -D VECT_TAB_OFFSET=$(bootldr_size)U
+GLOBAL_DEFINES += $(DEV_GLOBAL_DEFS) -D USE_RTOS_SYSTICK -D MODEL_$(model) -D VECT_TAB_OFFSET=$(bootldr_size)U
 
 # Without debug code
 DEBUG=
 ifdef swd_enable
-	GLOBAL_DEFINES += -DSWD_ENABLE
```
Comment (Ralim) [inline source/Makefile:134]: You appear to have removed the SWD_ENABLE flag?

**Comment by Ralim** [comment]
>I recommend not to execute make with a plain -j argument as it may slow down the build and instead specify an explicit value.

You are welcome to suggest an appropraite value to cover the range of 2-64 cores if you wish, I've just had `-j` work generally well.

>build.sh could be replaced by make clean all for each model.

Can you update build.sh to match as well?

Have you confirmed outputs are correct if you trigger builds for all languages at once?

**Review by alvinhochun — COMMENTED** [review]
> > I recommend not to execute make with a plain -j argument as it may slow down the build and instead specify an explicit value.
> 
> You are welcome to suggest an appropraite value to cover the range of 2-64 cores if you wish, I've just had `-j` work generally well.

The thing about specifying `-j` with no argument is that it runs as many jobs as possible without limiting to the number of cores, which sometimes makes the build *slower*. With this PR the final linking step for each languages are independent with each other, which means they can all run in parallel. Running 25 linker processes at the same time can really hog the CPU.

On Linux, you can probably use `-j $(grep -c ^processor /proc/cpuinfo)` as a default value (thanks StackOverflow).

File: source/Makefile [hunk 3]
```diff
@@ -275,49 +267,56 @@ OBJS_S = $(S_SRCS:.S=.o)
...
 	@$(AS) -c $(AFLAGS) $< -o $@
 
-$(APP_INC_DIR)/unit.h $(GENERATED_TRANSLATION_SOURCE): $(TRANSLATION_FILES) Makefile ../Translations/make_translation.py
-	@echo 'Building translations'
-	@cd ../Translations && python3 make_translation.py
-
+Core/Gen/Translation.%.cpp: ../Translations/translation_%.json Makefile ../Translations/make_translation.py ../Translations/translations_commons.js
+	@test -d $(@D) || mkdir -p $(@D)
+	@echo 'Generating translations for language $*'
+	@python3 ../Translations/make_translation.py $* > $(PWD)/$@ 2>/dev/null
```
[the remark this answers is not in the document]
Comment (t-8ch) [reply source/Makefile:316]: @alvinhochun Yep, I'll both of it.
I'll problably trim down the default output as the parallel execution may lead to interspersed, confusing messages.

**Comment by t-8ch** [comment]
For the correct number of jobs on Linux the best way to compute the value by using `make -j$(nproc)`, which uses the correct APIs to figure out the number of cores available.

On windows the `nproc` utility may also be available and functional, depending on how make and bash are provided.
Also there seems to be an environment variable `NUMBER_OF_PROCESSORS` on windows that can be used for that.
I don't know which setups are supported on windows and I don't have a windows machine to test, I can't validate the functionality there.

File: source/Makefile (hunk 2 above)
Comment (t-8ch) [reply source/Makefile:134]: Indeed, will readd.

File: source/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (alvinhochun) [reply source/Makefile:316]: > I'll problably trim down the default output as the parallel execution may lead to interspersed, confusing messages.

`make --output-sync=target` should fix that, but I think making sure that all output lines mentions the language code is also a good idea if it's not too troublesome.

**Comment by Ralim** [comment]
As far as i'm concerned windows is only supported under WSL to keep life sane; so `$(nproc)` is fine.
I agree it can be stressing; maybe thats just my luck of having high core count processors for too long. Absolutely fine for it to be changed.

File: source/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ralim) [reply source/Makefile:316]: I would prefer any changes to make things _more_ obvious in CI where possible.

**Comment by t-8ch** [comment]
Requested changes should be incorporated. After the final review I'd like to squash/reorganize the commits before merging.

**Comment by t-8ch** [comment]
Testing looks good. The firmwares are built with the correct language and Fahrenheit flags.

**Review by Ralim — APPROVED** [review]

**Comment by Ralim** [comment]
I'm happy with all of this I believe; so if you want to cleanup / finish up and let me know I'm happy to merge.
I can squash merge to keep history clean if you prefer.

**Comment by t-8ch** [comment]
@Ralim  will do!

FYI when I added `--output-sync` to the `make` invocation then the GitHub builds just got stuck completely. No clue why.
Maybe you can check if the current output is fine.

**Comment by t-8ch** [comment]
@Ralim  the commits are now squashed and the commit message change to properly reflect the final state of this PR.

FYI a quick check indicates that this PR reduced the build-step during CI by about 75%, total time by around 50%.

**Comment by Ralim** [comment]
I'll do a test on all of the units I have here tonight and if it looks good I'll merge it in

**Comment by Ralim** [comment]
If you could remerge master into this branch would be fantastic, but I can if you don't have time today

**Comment by t-8ch** [comment]
@Ralim I rebased onto current master

**Comment by Ralim** [comment]
❤️ 🤟🏼

**Comment by Ralim** [comment]
Thank you so much for all of the effort you have put into this, I massively appreciate the effort and the improvements :)

**Pull request merged**